packages/eslint-plugin/src/rules/no-array-delete.ts

```diff
@@ -0,0 +1,92 @@
+import type { TSESLint, TSESTree } from '@typescript-eslint/utils';
+import { AST_NODE_TYPES } from '@typescript-eslint/utils';
+import { unionTypeParts } from 'tsutils';
+import * as ts from 'typescript';
+
+import * as util from '../util';
+
+export default util.createRule({
+  name: 'no-array-delete',
+  meta: {
+    hasSuggestions: true,
+    docs: {
+      description: 'Disallow delete operator for arrays',
+      recommended: 'strict',
+      requiresTypeChecking: true,
+    },
+    messages: {
+      arrayDelete: 'Using the delete operator on an array is dangerous.',
+    },
+    schema: [],
+    type: 'problem',
+    fixable: 'code',
+  },
+  defaultOptions: [],
+  create(context) {
+    return {
+      "UnaryExpression[operator='delete'] > MemberExpression[computed]"(
+        node: TSESTree.MemberExpressionComputedName & {
+          parent: TSESTree.UnaryExpression;
+        },
+      ): void {
+        const parserServices = util.getParserServices(context);
+        const checker = parserServices.program.getTypeChecker();
+        const originalNode = parserServices.esTreeNodeToTSNodeMap.get(node);
+
+        const target = originalNode.getChildAt(0);
+        const key = originalNode.getChildAt(2);
+
+        const targetType = util.getConstrainedTypeAtLocation(checker, target);
+
+        if (!util.isTypeAnyArrayType) {
+          return;
+        }
+
+        const keyType = util.getConstrainedTypeAtLocation(checker, key);
+
+        if (
+          !isTypeArrayTypeOrArrayInUnionOfTypes(targetType, checker) ||
+          !isTypeNumberOrNumberLiteralOrNumberLikeType(keyType)
+        ) {
+          return;
+        }
```

Comment:
can we split this condition to exit early?

```suggestion
        if (!isTypeArrayTypeOrArrayInUnionOfTypes(targetType, checker)) {
          return;
        }

        const keyType = util.getConstrainedTypeAtLocation(checker, key);

        if (!isTypeNumberOrNumberLiteralOrNumberLikeType(keyType)) {
          return;
        }
```

in best case scenario this should reduce amount of type requests

Reply:
Thank you ❤️, didn't see that